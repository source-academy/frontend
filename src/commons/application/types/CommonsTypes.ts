import { Router } from '@remix-run/router';

export const LOG_OUT = 'LOG_OUT';
export const UPDATE_REACT_ROUTER = 'UPDATE_REACT_ROUTER';

export type RouterState = Router | null;
