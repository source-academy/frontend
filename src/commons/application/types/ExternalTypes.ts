/**
 * Defined for displaying an external library.
 * @see Library the definition of a Library at AssessmentTypes.ts in an assessment.
 */
export type External = {
  key: number;
  name: ExternalLibraryName;
  symbols: string[];
};

export enum ExternalLibraryName {
  NONE = 'NONE',
  SOUNDS = 'SOUNDS',
  EV3 = 'EV3' // Remote execution device library
}

export type ExternalLibrary = {
  name: ExternalLibraryName;
  symbols: string[];
};

/**
 * Defines which external libraries are available for usage, and what
 * external symbols (exposed functions) are under them.
 */

const libEntries: Array<[ExternalLibraryName, string[]]> = [
  [ExternalLibraryName.NONE, []],
  [ExternalLibraryName.SOUNDS, []]
];

export const externalLibraries: Map<string, string[]> = new Map(libEntries);
