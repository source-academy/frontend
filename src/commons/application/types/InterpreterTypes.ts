export const BEGIN_INTERRUPT_EXECUTION = 'BEGIN_INTERRUPT_EXECUTION';
export const END_INTERRUPT_EXECUTION = 'END_INTERRUPT_EXECUTION';
export const EVAL_INTERPRETER_ERROR = 'EVAL_INTERPRETER_ERROR';
export const EVAL_INTERPRETER_SUCCESS = 'EVAL_INTERPRETER_SUCCESS';
export const EVAL_TESTCASE_FAILURE = 'EVAL_TESTCASE_FAILURE';
export const EVAL_TESTCASE_SUCCESS = 'EVAL_TESTCASE_SUCCESS';
export const HANDLE_CONSOLE_LOG = 'HANDLE_CONSOLE_LOG';
export const BEGIN_DEBUG_PAUSE = 'BEGIN_DEBUG_PAUSE';
export const END_DEBUG_PAUSE = 'END_DEBUG_PAUSE';
export const DEBUG_RESUME = 'DEBUG_RESUME';
export const DEBUG_RESET = 'DEBUG_RESET';
export const UPDATE_EDITOR_HIGHLIGHTED_LINES = 'UPDATE_EDITOR_HIGHLIGHTED_LINES';
