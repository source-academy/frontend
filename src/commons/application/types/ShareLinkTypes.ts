export type ShareLinkShortenedUrlResponse = {
  shortenedUrl: string;
};
