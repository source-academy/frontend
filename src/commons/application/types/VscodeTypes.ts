export type VscodeState = {
  isVscode: boolean;
};
