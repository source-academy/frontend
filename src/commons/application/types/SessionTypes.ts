import { Octokit } from '@octokit/rest';
import { Chapter, Variant } from 'js-slang/dist/types';

import { Grading, GradingOverview } from '../../../features/grading/GradingTypes';
import { Device, DeviceSession } from '../../../features/remoteExecution/RemoteExecutionTypes';
import {
  Assessment,
  AssessmentConfiguration,
  AssessmentOverview
} from '../../assessment/AssessmentTypes';
import { MissionRepoData } from '../../githubAssessments/GitHubMissionTypes';
import { Notification } from '../../notificationBadge/NotificationBadgeTypes';
import { GameState, Role, Story } from '../ApplicationTypes';

export const FETCH_AUTH = 'FETCH_AUTH';
export const FETCH_USER_AND_COURSE = 'FETCH_USER_AND_COURSE';
export const FETCH_COURSE_CONFIG = 'FETCH_COURSE_CONFIG';
export const FETCH_ASSESSMENT = 'FETCH_ASSESSMENT';
export const FETCH_ASSESSMENT_ADMIN = 'FETCH_ASSESSMENT_ADMIN';
export const FETCH_ASSESSMENT_OVERVIEWS = 'FETCH_ASSESSMENT_OVERVIEWS';
export const FETCH_ALL_USER_XP = 'FETCH_ALL_USER_XP';
export const FETCH_TOTAL_XP = 'FETCH_TOTAL_XP';
export const FETCH_TOTAL_XP_ADMIN = 'FETCH_TOTAL_XP_ADMIN';
export const FETCH_GRADING = 'FETCH_GRADING';
export const FETCH_GRADING_OVERVIEWS = 'FETCH_GRADING_OVERVIEWS';
export const LOGIN = 'LOGIN';
export const LOGOUT_GOOGLE = 'LOGOUT_GOOGLE';
export const LOGIN_GITHUB = 'LOGIN_GITHUB';
export const LOGOUT_GITHUB = 'LOGOUT_GITHUB';
export const SET_TOKENS = 'SET_TOKENS';
export const SET_USER = 'SET_USER';
export const SET_COURSE_CONFIGURATION = 'SET_COURSE_CONFIGURATION';
export const SET_COURSE_REGISTRATION = 'SET_COURSE_REGISTRATION';
export const SET_ASSESSMENT_CONFIGURATIONS = 'SET_ASSESSMENT_CONFIGURATIONS';
export const SET_ADMIN_PANEL_COURSE_REGISTRATIONS = 'SET_ADMIN_PANEL_COURSE_REGISTRATIONS';
export const SET_GOOGLE_USER = 'SET_GOOGLE_USER';
export const SET_GITHUB_ASSESSMENT = 'SET_GITHUB_ASSESSMENT';
export const SET_GITHUB_OCTOKIT_OBJECT = 'SET_GITHUB_OCTOKIT_OBJECT';
export const SET_GITHUB_ACCESS_TOKEN = 'SET_GITHUB_ACCESS_TOKEN';
export const SUBMIT_ANSWER = 'SUBMIT_ANSWER';
export const SUBMIT_ASSESSMENT = 'SUBMIT_ASSESSMENT';
export const SUBMIT_GRADING = 'SUBMIT_GRADING';
export const SUBMIT_GRADING_AND_CONTINUE = 'SUBMIT_GRADING_AND_CONTINUE';
export const REAUTOGRADE_SUBMISSION = 'REAUTOGRADE_SUBMISSION';
export const REAUTOGRADE_ANSWER = 'REAUTOGRADE_ANSWER';
export const REMOVE_GITHUB_OCTOKIT_OBJECT_AND_ACCESS_TOKEN =
  'REMOVE_GITHUB_OCTOKIT_OBJECT_AND_ACCESS_TOKEN';
export const UNSUBMIT_SUBMISSION = 'UNSUBMIT_SUBMISSION';
export const UPDATE_ASSESSMENT_OVERVIEWS = 'UPDATE_ASSESSMENT_OVERVIEWS';
export const UPDATE_TOTAL_XP = 'UPDATE_TOTAL_XP';
export const UPDATE_ALL_USER_XP = 'UPDATE_ALL_USER_XP';
export const UPDATE_ASSESSMENT = 'UPDATE_ASSESSMENT';
export const UPDATE_GRADING_OVERVIEWS = 'UPDATE_GRADING_OVERVIEWS';
export const UPDATE_GRADING = 'UPDATE_GRADING';
export const FETCH_NOTIFICATIONS = 'FETCH_NOTIFICATIONS';
export const ACKNOWLEDGE_NOTIFICATIONS = 'ACKNOWLEDGE_NOTIFICATIONS';
export const UPDATE_NOTIFICATIONS = 'UPDATE_NOTIFICATIONS';
export const UPDATE_LATEST_VIEWED_COURSE = 'UPDATE_LATEST_VIEWED_COURSE';
export const UPDATE_COURSE_CONFIG = 'UPDATE_COURSE_CONFIG';
export const FETCH_ASSESSMENT_CONFIGS = 'FETCH_ASSESSMENT_CONFIGS';
export const UPDATE_ASSESSMENT_CONFIGS = 'UPDATE_ASSESSMENT_CONFIGS';
export const DELETE_ASSESSMENT_CONFIG = 'DELETE_ASSESSMENT_CONFIG';
export const FETCH_ADMIN_PANEL_COURSE_REGISTRATIONS = 'FETCH_ADMIN_PANEL_COURSE_REGISTRATIONS';
export const UPDATE_USER_ROLE = 'UPDATE_USER_ROLE';
export const UPDATE_COURSE_RESEARCH_AGREEMENT = 'UPDATE_COURSE_RESEARCH_AGREEMENT';
export const DELETE_USER_COURSE_REGISTRATION = 'DELETE_USER_COURSE_REGISTRATION';

export const UPLOAD_KEYSTROKE_LOGS = 'UPLOAD_KEYSTROKE_LOGS';
export const UPLOAD_UNSENT_LOGS = 'UPLOAD_UNSENT_LOGS';

export type SessionState = {
  // Tokens
  readonly accessToken?: string;
  readonly refreshToken?: string;

  // User
  readonly userId?: number;
  readonly name?: string;
  readonly courses: UserCourse[];

  // Course Registration
  readonly courseRegId?: number;
  readonly role?: Role;
  readonly group: string | null;
  readonly gameState: GameState;
  readonly courseId?: number;
  readonly xp: number;
  readonly allUserXp: string[][] | undefined;
  readonly story: Story;

  // Course Configuration
  readonly courseName?: string;
  readonly courseShortName?: string;
  readonly viewable?: boolean;
  readonly enableGame?: boolean;
  readonly enableAchievements?: boolean;
  readonly enableSourcecast?: boolean;
  readonly sourceChapter?: number;
  readonly sourceVariant?: Variant;
  readonly moduleHelpText?: string;
  readonly assetsPrefix?: string;

  readonly assessmentConfigurations?: AssessmentConfiguration[];
  readonly userCourseRegistrations?: AdminPanelCourseRegistration[];

  // For research data collection
  readonly agreedToResearch?: boolean | null;
  readonly sessionId: number;

  readonly assessmentOverviews?: AssessmentOverview[];
  readonly assessments: Map<number, Assessment>;
  readonly gradingOverviews?: GradingOverview[];
  readonly gradings: Map<number, Grading>;
  readonly notifications: Notification[];
  readonly googleUser?: string;
  readonly githubAssessment?: MissionRepoData;
  readonly githubOctokitObject: { octokit: Octokit | undefined };
  readonly githubAccessToken?: string;
  readonly remoteExecutionDevices?: Device[];
  readonly remoteExecutionSession?: DeviceSession;
};

export type Tokens = {
  accessToken: string;
  refreshToken: string;
};

export type UserCourse = {
  courseId: number;
  courseName: string;
  courseShortName: string;
  role: Role;
  viewable: boolean;
};

export type User = {
  userId: number;
  name: string;
  courses: UserCourse[];
};

export type CourseRegistration = {
  courseRegId: number;
  role: Role;
  group: string | null;
  gameState?: GameState;
  courseId: number;
  xp: number;
  story?: Story;
  agreedToResearch: boolean | null;
};

export type CourseConfiguration = {
  courseName: string;
  courseShortName: string;
  viewable: boolean;
  enableGame: boolean;
  enableAchievements: boolean;
  enableSourcecast: boolean;
  sourceChapter: Chapter;
  sourceVariant: Variant;
  moduleHelpText: string;
  assetsPrefix: string;
};

export type AdminPanelCourseRegistration = {
  courseRegId: number;
  courseId: number;
  name?: string;
  username: string;
  role: Role;
  group?: string;
};

export type UpdateCourseConfiguration = Partial<CourseConfiguration>;
