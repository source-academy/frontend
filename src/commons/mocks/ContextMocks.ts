import { parse } from 'acorn';
import { FunctionExpression, Node } from 'estree';
import { ACORN_PARSE_OPTIONS } from 'js-slang/dist/constants';
import createContext, { EnvTree } from 'js-slang/dist/createContext';
import Closure from 'js-slang/dist/interpreter/closure';
import { Context, Environment } from 'js-slang/dist/types';
import { TypeError } from 'js-slang/dist/utils/rttc';

export function mockContext(chapter = 1): Context {
  return createContext(chapter);
}

export function mockRuntimeContext(): Context {
  const context = createContext();
  // Note: noticed no harm in removing the following context.runtime.
  // If you get an error with head is undefined after trying to evaluate code.
  // Likely due to the environments: [].
  // In every real context, there is at least one env frame.
  context.runtime = {
    break: false,
    debuggerOn: true,
    isRunning: true,
    environmentTree: new EnvTree(),
    environments: [],
    nodes: [
      {
        type: 'Literal',
        loc: {
          start: { line: 1, column: 0 },
          end: { line: 1, column: 1 }
        },
        value: 0,
        raw: '0',
        range: [0, 1]
      }
    ],
    agenda: null,
    stash: null,
    envStepsTotal: 0,
    breakpointSteps: []
  };
  return context;
}

export function mockClosure(): Closure {
  return new Closure({} as FunctionExpression, {} as Environment, {} as Context);
}

export function mockTypeError(): TypeError {
  // Typecast to Node to fix estree-acorn compatability.
  return new TypeError(parse('', ACORN_PARSE_OPTIONS) as Node, '', '', '');
}
