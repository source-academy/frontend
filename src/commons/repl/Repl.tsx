import { Card, Classes, Pre } from '@blueprintjs/core';
import classNames from 'classnames';
import { parseError } from 'js-slang';
import { Variant } from 'js-slang/dist/types';
import { stringify } from 'js-slang/dist/utils/stringify';
import * as React from 'react';
import AceEditor from 'react-ace';
import { HotKeys } from 'react-hotkeys';

import { InterpreterOutput } from '../application/ApplicationTypes';
import { ExternalLibraryName } from '../application/types/ExternalTypes';
import SideContentCanvasOutput from '../sideContent/SideContentCanvasOutput';
import { ReplInput } from './ReplInput';
import { OutputProps } from './ReplTypes';

export type ReplProps = DispatchProps & StateProps & OwnProps;

type StateProps = {
  output: InterpreterOutput[];
  replValue: string;
  hidden?: boolean;
  inputHidden?: boolean;
  usingSubst?: boolean;
  sourceChapter: number;
  sourceVariant: Variant;
  externalLibrary: ExternalLibraryName;
  disableScrolling?: boolean;
};

type DispatchProps = {
  handleBrowseHistoryDown: () => void;
  handleBrowseHistoryUp: () => void;
  handleReplEval: () => void;
  handleReplValueChange: (newCode: string) => void;
};

type OwnProps = {
  replButtons: Array<JSX.Element | null>;
};

const Repl = React.forwardRef<AceEditor, ReplProps>((props, ref) => {
  const cards = props.output.map((slice, index) => (
    <Output output={slice} key={index} usingSubst={props.usingSubst || false} />
  ));
  return (
    <div className="Repl" style={{ display: props.hidden ? 'none' : undefined }}>
      <div className="repl-output-parent">
        {cards}
        {!props.inputHidden && (
          <HotKeys
            className={classNames('repl-input-parent', 'row', Classes.CARD, Classes.ELEVATION_0)}
            handlers={handlers}
          >
            <ReplInput {...props} ref={ref} />
          </HotKeys>
        )}
      </div>
    </div>
  );
});

export const Output: React.FC<OutputProps> = (props: OutputProps) => {
  switch (props.output.type) {
    case 'code':
      return (
        <Card>
          <Pre className="codeOutput">{props.output.value}</Pre>
        </Card>
      );
    case 'running':
      return (
        <Card>
          <Pre className="logOutput">{props.output.consoleLogs.join('\n')}</Pre>
        </Card>
      );
    case 'result':
      // We check if we are using Stepper, so we can process the REPL results properly
      if (props.usingSubst && props.output.value instanceof Array) {
        return (
          <Card>
            <Pre className="logOutput">Check out the Stepper tab!</Pre>
          </Card>
        );
      } else if (props.output.consoleLogs.length === 0) {
        return (
          <Card>
            <Pre className="resultOutput">{renderResult(props.output.value)}</Pre>
          </Card>
        );
      } else {
        return (
          <Card>
            <Pre className="logOutput">{props.output.consoleLogs.join('\n')}</Pre>
            <Pre className="resultOutput">{renderResult(props.output.value)}</Pre>
          </Card>
        );
      }
    case 'errors':
      if (props.output.consoleLogs.length === 0) {
        return (
          <Card>
            <Pre className="errorOutput">{parseError(props.output.errors)}</Pre>
          </Card>
        );
      } else {
        return (
          <Card>
            <Pre className="logOutput">{props.output.consoleLogs.join('\n')}</Pre>
            <br />
            <Pre className="errorOutput">{parseError(props.output.errors)}</Pre>
          </Card>
        );
      }
    default:
      return <Card>''</Card>;
  }
};

const renderResult = (value: any) => {
  /** A class which is the output of the show() function */
  const ShapeDrawn = (window as any).ShapeDrawn;
  if (typeof ShapeDrawn !== 'undefined' && value instanceof ShapeDrawn) {
    return <SideContentCanvasOutput canvas={value.$canvas} />;
  } else {
    return stringify(value);
  }
};

/* Override handler, so does not trigger when focus is in editor */
const handlers = {
  goGreen: () => {}
};

export default Repl;
