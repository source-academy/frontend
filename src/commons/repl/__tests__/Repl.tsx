import { shallow } from 'enzyme';
import { Chapter, Variant } from 'js-slang/dist/types';
import { ExternalLibraryName } from 'src/commons/application/types/ExternalTypes';

import {
  CodeOutput,
  ErrorOutput,
  InterpreterOutput,
  ResultOutput,
  RunningOutput
} from '../../application/ApplicationTypes';
import { mockTypeError } from '../../mocks/ContextMocks';
import Repl, { Output } from '../Repl';

const mockRunningOutput: RunningOutput = {
  type: 'running',
  consoleLogs: ['a', 'bb', 'cccccccccccccccccccccccccccccccc', 'd']
};

const mockCodeOutput: CodeOutput = {
  type: 'code',
  value: "display('');"
};

const mockResultOutput: ResultOutput = {
  type: 'result',
  value: 42,
  consoleLogs: []
};

const mockErrorOutput: ErrorOutput = {
  type: 'errors',
  errors: [mockTypeError()],
  consoleLogs: []
};

test('Repl renders correctly', () => {
  const props = {
    handleBrowseHistoryDown: () => {},
    handleBrowseHistoryUp: () => {},
    handleReplValueChange: (newCode: string) => {},
    handleReplEval: () => {},
    handleReplOutputClear: () => {},
    output: [mockResultOutput, mockCodeOutput, mockErrorOutput, mockRunningOutput],
    replValue: '',
    sourceChapter: Chapter.SOURCE_1,
    sourceVariant: Variant.DEFAULT,
    externalLibrary: ExternalLibraryName.NONE,
    replButtons: []
  };
  const app = <Repl {...props} />;
  const tree = shallow(app);
  expect(tree.debug()).toMatchSnapshot();
});

test('Code output renders correctly', () => {
  const app = <Output {...{ output: mockCodeOutput }} />;
  const tree = shallow(app);
  expect(tree.debug()).toMatchSnapshot();
});

test('Running output renders correctly', () => {
  const app = <Output {...{ output: mockRunningOutput }} />;
  const tree = shallow(app);
  expect(tree.debug()).toMatchSnapshot();
});

test('Result output (no consoleLogs) renders correctly', () => {
  const app = <Output {...{ output: mockResultOutput }} />;
  const tree = shallow(app);
  expect(tree.debug()).toMatchSnapshot();
});

test('Result output (with consoleLogs) renders correctly', () => {
  const props = {
    ...mockResultOutput,
    consoleLogs: mockRunningOutput.consoleLogs
  };
  const app = <Output {...{ output: props }} />;
  const tree = shallow(app);
  expect(tree.debug()).toMatchSnapshot();
});

test('Error output (no consoleLogs) renders correctly', () => {
  const app = <Output {...{ output: mockErrorOutput }} />;
  const tree = shallow(app);
  expect(tree.debug()).toMatchSnapshot();
});

test('Error output (with consoleLogs) renders correctly', () => {
  const props = {
    ...mockErrorOutput,
    consoleLogs: mockRunningOutput.consoleLogs
  };
  const app = <Output {...{ output: props }} />;
  const tree = shallow(app);
  expect(tree.debug()).toMatchSnapshot();
});

test('Empty output renders an empty card', () => {
  const app = <Output {...{ output: {} as InterpreterOutput }} />;
  const tree = shallow(app);
  expect(tree.debug()).toMatchSnapshot();
});
