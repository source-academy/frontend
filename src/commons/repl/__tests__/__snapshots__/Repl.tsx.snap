// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Code output renders correctly 1`] = `
"<Blueprint4.Card elevation={0} interactive={false}>
  <Component className=\\"code-output\\">
    display(&#39;&#39;);
  </Component>
</Blueprint4.Card>"
`;

exports[`Empty output renders an empty card 1`] = `
"<Blueprint4.Card elevation={0} interactive={false}>
  &#39;&#39;
</Blueprint4.Card>"
`;

exports[`Error output (no consoleLogs) renders correctly 1`] = `
"<Blueprint4.Card elevation={0} interactive={false}>
  <Component className=\\"error-output\\">
    Expected , got .
  </Component>
</Blueprint4.Card>"
`;

exports[`Error output (with consoleLogs) renders correctly 1`] = `
"<Blueprint4.Card elevation={0} interactive={false}>
  <Component className=\\"log-output\\">
    a
    bb
    cccccccccccccccccccccccccccccccc
    d
  </Component>
  <br />
  <Component className=\\"error-output\\">
    Expected , got .
  </Component>
</Blueprint4.Card>"
`;

exports[`Repl renders correctly 1`] = `
"<div className=\\"Repl\\" style={{...}}>
  <div className=\\"repl-output-parent\\">
    <Output output={{...}} usingSubst={false} isHtml={false} />
    <Output output={{...}} usingSubst={false} isHtml={false} />
    <Output output={{...}} usingSubst={false} isHtml={false} />
    <Output output={{...}} usingSubst={false} isHtml={false} />
    <HotKeysEnabled className=\\"repl-input-parent row bp4-card bp4-elevation-0\\" handlers={{...}}>
      <ReplInput handleBrowseHistoryDown={[Function: handleBrowseHistoryDown]} handleBrowseHistoryUp={[Function: handleBrowseHistoryUp]} handleReplValueChange={[Function: handleReplValueChange]} handleReplEval={[Function: handleReplEval]} handleReplOutputClear={[Function: handleReplOutputClear]} output={{...}} replValue=\\"\\" sourceChapter={1} sourceVariant=\\"default\\" externalLibrary=\\"NONE\\" replButtons={{...}} />
    </HotKeysEnabled>
  </div>
</div>"
`;

exports[`Result output (no consoleLogs) renders correctly 1`] = `
"<Blueprint4.Card elevation={0} interactive={false}>
  <Component className=\\"result-output\\">
    42
  </Component>
</Blueprint4.Card>"
`;

exports[`Result output (with consoleLogs) renders correctly 1`] = `
"<Blueprint4.Card elevation={0} interactive={false}>
  <Component className=\\"log-output\\">
    a
    bb
    cccccccccccccccccccccccccccccccc
    d
  </Component>
  <Component className=\\"result-output\\">
    42
  </Component>
</Blueprint4.Card>"
`;

exports[`Running output renders correctly 1`] = `
"<Blueprint4.Card elevation={0} interactive={false}>
  <Component className=\\"log-output\\">
    a
    bb
    cccccccccccccccccccccccccccccccc
    d
  </Component>
</Blueprint4.Card>"
`;
